src/index.tsx

```diff
@@ -61,11 +72,74 @@ export type Elements =
   | 'image_caption'
   | 'quote';
 
+type CustomText = { text: string; children: [] } & { [key in Mark]?: boolean };
+
+type ImageElement = {
+  type: 'image';
+  src: string;
+  objectUrl: string;
+  children: [];
+};
+type ImageCaptionElement = { type: 'image_caption'; children: CustomText[] };
+type FigureElement = {
+  type: 'figure';
+  children: (ImageElement | ImageCaptionElement)[];
+};
+type TextElement = {
+  type: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'paragraph';
+  children: CustomText[];
+};
+type CodeBlockElement = {
+  type: 'code_block';
+  children: TextElement[];
+};
+
+type QuoteElement = {
+  type: 'quote';
+  children: CustomText[];
+};
+type ListItemElement = {
+  type: 'list_item';
+  children: TextElement[];
+};
+type ListElement = {
+  type: 'ol_list' | 'ul_list';
+  children: (ListElement | ListItemElement)[];
+};
+type LinkElement = {
+  type: 'link';
+  children: (TextElement | CustomText)[];
+  url: string;
+};
+
+type CustomElement =
+  | ListElement
+  | ListItemElement
+  | TextElement
+  | ImageElement
+  | ImageCaptionElement
+  | FigureElement
+  | QuoteElement
+  | CodeBlockElement
+  | LinkElement;
+
+declare module 'slate' {
+  interface CustomTypes {
+    Editor: BaseEditor &
+      ReactEditor &
+      HistoryEditor &
+      ListEditor &
+      PluginsEditor;
+    Element: CustomElement;
+    Text: CustomText;
+  }
+}
+
```

Comment:
This file is getting quite huge, so I'll probably move all of the type stuff into a separate file :grimacing: